src/Microsoft.Identity.Client/Internal/ClientCredentialWrapper.cs

```diff
@@ -17,41 +17,51 @@ namespace Microsoft.Identity.Client.Internal
     /// </summary>
     internal sealed class ClientCredentialWrapper
     {
-        /// <summary>
-        /// Constructor of client (application) credentials from a <see cref="ClientAssertionCertificateWrapper"/>
-        /// </summary>
-        /// <param name="certificate">contains information about the certificate previously shared with AAD at application
-        /// registration to prove the identity of the application (the client) requesting the tokens.</param>
-        public ClientCredentialWrapper(ClientAssertionCertificateWrapper certificate)
+        public static ClientCredentialWrapper CreateWithCertificate(ClientAssertionCertificateWrapper certificate)
+        {
+           return new ClientCredentialWrapper(certificate);
+        }
+        public static ClientCredentialWrapper CreateWithSecret(string secret)
+        {
+            return new ClientCredentialWrapper(secret, false);
+        }
+        public static ClientCredentialWrapper CreateWithSignedClientAssertion(string signedClientAssertion)
+        {
+            return new ClientCredentialWrapper(signedClientAssertion, true);
+        }
+
+        private ClientCredentialWrapper(ClientAssertionCertificateWrapper certificate)
         {
             ConfidentialClientApplication.GuardMobileFrameworks();
             Certificate = certificate;
         }
 
-        internal ClientAssertionCertificateWrapper Certificate { get; private set; }
-        internal string Assertion { get; set; }
-        internal long ValidTo { get; set; }
-        internal bool ContainsX5C { get; set; }
-        internal string Audience { get; set; }
-
-        /// <summary>
-        /// Constructor of client (application) credentials from a client secret, also known as the application password.
-        /// </summary>
-        /// <param name="secret">Secret string previously shared with AAD at application registration to prove the identity
-        /// of the application (the client) requesting the tokens.</param>
-        public ClientCredentialWrapper(string secret)
+        private ClientCredentialWrapper(string secretOrAssertion, bool isSignedAssertion)
```

Comment:
Adding factory methods here to avoid having 2 constructors with the same parameter type (string) Since this method can take either a signed assertion or a secret, I called the string "secretOrAssertion". Any other suggestions on how to resolve this? @henrik-me @bgavrilMS

Reply:
Static builders is probably the simplest.